### PR TITLE
[PATCH API-NEXT v2] api: ipsec: improve salt specification

### PR DESCRIPTION
**Pull request by psavol**
Both crypto and IPSEC API documentation needed improvement on how key, IV and extra keying material paramters are expected to be used. Especially, AEAD algorithms needed additional specification. Also, IPSEC with GMAC need a new parameter for extra auth key material.

### CODE REVIEW

File: include/odp/api/spec/crypto.h [hunk 1]
```diff
@@ -435,18 +438,31 @@ typedef struct odp_crypto_session_param_t {
...
 	odp_crypto_op_mode_t op_mode;
 
 	/** Cipher algorithm
 	 *
-	 *  Use odp_crypto_capability() for supported algorithms.
+	 *  Select cipher algorithm to be used. ODP_CIPHER_ALG_NULL indicates
+	 *  that ciphering is disabled. Use odp_crypto_capability() for
+	 *  supported algorithms. Note that some algorithms restrict choice of
+	 *  the pairing authentication algorithm. When ciphering is enabled
+	 *  cipher key and IV needs to be set. The default value is
```
Comment (Bill-Fischofer-Linaro) [inline include/odp/api/spec/crypto.h]: Grammar (agreement): `need to be set`

File: include/odp/api/spec/crypto.h [hunk 2]
```diff
@@ -467,7 +483,18 @@ typedef struct odp_crypto_session_param_t {
...
+	 *  Select authentication algorithm to be used. ODP_AUTH_ALG_NULL
+	 *  indicates that authentication is disabled. Use
+	 *  odp_crypto_capability() for supported algorithms. Note that some
+	 *  algorithms restrict choice of the pairing cipher algorithm. When
+	 *  single algorithm provides both ciphering and authentication
+	 *  (i.e. Authenticated Encryption), authentication side key
+	 *  (auth_key) and IV (auth_iv) are ignored, and cipher side values are
+	 *  used instead. These algorithms ignore authentication side key
+	 *  and IV: ODP_AUTH_ALG_AES_GCM, ODP_AUTH_ALG_AES_CCM and
+	 *  ODP_AUTH_ALG_CHACHA20_POLY1305. Otherwise, all authentication side
```
Comment (Bill-Fischofer-Linaro) [inline include/odp/api/spec/crypto.h]: Do we want/need to enumerate these here? Each implementation chooses which algorithms it implements (as indicated by `odp_crypto_capability()`. It's up to each user of those algorithms to understand their associated usage conventions.

File: include/odp/api/spec/ipsec.h [hunk 3]
```diff
@@ -364,27 +364,65 @@ typedef enum odp_ipsec_tunnel_type_t {
...
-	/** Cipher algorithm */
+	/** Cipher algorithm
+	 *
+	 *  Select cipher algorithm to be used. ODP_CIPHER_ALG_NULL indicates
+	 *  that ciphering is disabled. See 'ciphers' field of
+	 *  odp_ipsec_capability_t for supported cipher algorithms. Algorithm
+	 *  descriptions can be found from odp_cipher_alg_t documentation. Note
+	 *  that some algorithms restrict choice of the pairing authentication
+	 *  algorithm. When ciphering is enabled, cipher key and potential extra
+	 *  key material (cipher_key_extra) needs to be set. The default value
```
Comment (Bill-Fischofer-Linaro) [inline include/odp/api/spec/ipsec.h]: `need to be set`

**Comment by Bill-Fischofer-Linaro** [comment]
@lumag please review this PR as well.

File: test/validation/api/ipsec/ipsec.c [hunk 4]
```diff
@@ -378,8 +378,15 @@ void ipsec_sa_param_fill(odp_ipsec_sa_param_t *param,
 	if (auth_key)
 		param->crypto.auth_key = *auth_key;
 
-	if (extra_key)
-		param->crypto.cipher_key_extra = *extra_key;
```
Comment (lumag) [inline test/validation/api/ipsec/ipsec.c]: I think it would be more logical to introduce `extra_cipher_key`/`extra_auth_key` here.

File: test/performance/odp_ipsec.c [hunk 5]
```diff
@@ -277,6 +277,10 @@ static ipsec_alg_config_t algs_config[] = {
 				.data = test_key16,
 				.length = sizeof(test_key16)
 			},
+			.cipher_key_extra = {
+				.data = test_salt,
+				.length = 4,
+			},
```
Comment (lumag) [inline test/performance/odp_ipsec.c]: Could you please separate these two hunks into separate commit: I'd like to get them into master branch as well.

File: platform/linux-generic/odp_ipsec_sad.c [hunk 6]
```diff
@@ -523,17 +541,19 @@ odp_ipsec_sa_t odp_ipsec_sa_create(const odp_ipsec_sa_param_t *param)
...
 	ipsec_sa->icv_len = crypto_param.auth_digest_len;
 
-	if (param->crypto.cipher_key_extra.length) {
-		if (param->crypto.cipher_key_extra.length >
-		    IPSEC_MAX_SALT_LEN)
+	if (ipsec_sa->salt_length) {
+		if (ipsec_sa->salt_length > IPSEC_MAX_SALT_LEN) {
+			ODP_ERR("IPSEC_MAX_SALT_LEN too small\n");
 			goto error;
+		}
```
Comment (lumag) [inline platform/linux-generic/odp_ipsec_sad.c]: Argh. This needs to be fixed in master branch as well.

File: include/odp/api/spec/crypto.h (hunk 1 above)
Comment (psavol) [reply include/odp/api/spec/crypto.h]: ok

File: include/odp/api/spec/crypto.h (hunk 2 above)
Comment (psavol) [reply include/odp/api/spec/crypto.h]: These enumerated values change the default usage of API. With all other auth algos, user must fill in all auth parameters in this struct. When auth algo is one of these three, auth_key and auth_iv params are ignored and cipher ones are used instead (as those would be the same anyway).

File: include/odp/api/spec/ipsec.h (hunk 3 above)
Comment (psavol) [reply include/odp/api/spec/ipsec.h]: ok

File: test/validation/api/ipsec/ipsec.c (hunk 4 above)
Comment (psavol) [reply test/validation/api/ipsec/ipsec.c]: This was minimal change to test cases, but agree test cases are easier to read when those are separate params. Done in v2.

File: test/performance/odp_ipsec.c (hunk 5 above)
Comment (psavol) [reply test/performance/odp_ipsec.c]: Done in v2

File: platform/linux-generic/odp_ipsec_sad.c (hunk 6 above)
Comment (psavol) [reply platform/linux-generic/odp_ipsec_sad.c]: Done in v2.

**Comment by muvarov** [comment]
Merged.
